### PR TITLE
Improve sampling for variogram estimation: circular and ring sampling + tests

### PR DESCRIPTION
**Pull request by rhugonnet**
Added ring/circular shape masking to subsample rasters.
This will extremely useful to better estimate empirical variograms from the raster samples. Because we generally have millions of samples in a raster, we can't calculate the millions^{squared} pairwise differences. Sampling 10,000 points is already quite a long calculation (~1min). If we subsample within rings of different radius, we limit the samples to similar pairwise distances, thus improving the empirical estimation with less unused samples.
TODO: write the parent function to subsample with the ring method, iteratively, for different radius sizes.

### CODE REVIEW

**Comment by erikmannerfelt** [comment]
Nice! Especially the `verbose` setting ;) 

Impressively old branch haha! 129 commits behind main!

**Comment by rhugonnet** [comment]
@erikmannerfelt Any idea why mamba installation fails? I don't get it, nothing should have changed on this

**Comment by erikmannerfelt** [comment]
@rhugonnet, if you check the log, it says "Connection broken" on that step. It's a bug with GitHub (not related to xdem) where it sometimes just randomly fails... When that happens, you have to trigger a re-run of all jobs!

I just triggered a re-run.

File: tests/test_spstats.py [hunk 1]
```diff
@@ -18,11 +19,10 @@
     warnings.filterwarnings("ignore", category=DeprecationWarning)
     from skgstat import models
 
-
 PLOT = False
 
-
-def load_diff() -> tuple[gu.georaster.Raster, np.ndarray]:
+def load_diff() -> Tuple[gu.georaster.Raster, np.ndarray] :
```
Comment (erikmannerfelt) [inline tests/test_spstats.py:24]: Why did you change this from `tuple` to `Tuple`? If a linter suggests it, the linter is outdated.

Comment (rhugonnet) [reply tests/test_spstats.py:24]: I does this to me for `list` and `tuple`, and wants me to import `List` and `Tuple` from annotations instead. Are those two similarly outdated?

**Comment by rhugonnet** [comment]
> @rhugonnet, if you check the log, it says "Connection broken" on that step. It's a bug with GitHub (not related to xdem) where it sometimes just randomly fails... When that happens, you have to trigger a re-run of all jobs!
> 
> I just triggered a re-run.

Thanks, good to know! Didn't see it on the log (I did not think it could happen actually).

File: xdem/spstats.py [hunk 2]
```diff
@@ -91,6 +90,73 @@ def random_subset(dh: np.ndarray, coords: np.ndarray, nsamp: int):
 
     return dh_sub, coords_sub
 
+def create_circular_mask(h: float, w: float, center: Optional[list[float]] = None, radius: Optional[float] = None) -> np.ndarray:
+    """
+    Create circular mask on a raster
+
+    :param h: height of array
```
Comment (adehecq) [inline xdem/spstats.py:97]: Maybe use a shape directly? It's simpler as one can use `array.shape` as input instead.

File: xdem/spstats.py [hunk 3]
```diff
@@ -91,6 +90,73 @@ def random_subset(dh: np.ndarray, coords: np.ndarray, nsamp: int):
 
     return dh_sub, coords_sub
 
+def create_circular_mask(h: float, w: float, center: Optional[list[float]] = None, radius: Optional[float] = None) -> np.ndarray:
```
Comment (adehecq) [inline xdem/spstats.py:93]: Just found out there is a skimage function that does just that! https://www.kite.com/python/docs/skimage.draw.circle

File: xdem/spstats.py [hunk 4]
```diff
@@ -91,6 +90,73 @@ def random_subset(dh: np.ndarray, coords: np.ndarray, nsamp: int):
...
+    """
+    Create circular mask on a raster
+
+    :param h: height of array
+    :param w: width of array
+    :param center: center
```
Comment (adehecq) [inline xdem/spstats.py:99]: Good to mention what the default will be, as it's unclear from the function definition.

File: xdem/spstats.py [hunk 5]
```diff
@@ -91,6 +90,73 @@ def random_subset(dh: np.ndarray, coords: np.ndarray, nsamp: int):
...
+
+    mask = dist_from_center <= radius
+
+    return mask
+
+def create_ring_mask(h: float, w: float, center: Optional[list[float]] = None, in_radius: float = 0., out_radius: float = 0.) -> np.ndarray:
+    """
+    Create ring mask on a raster
+
+    :param h: height of array
```
Comment (adehecq) [inline xdem/spstats.py:120]: Same points as for the previous functions about the arguments.

File: xdem/spstats.py [hunk 6]
```diff
@@ -91,6 +90,73 @@ def random_subset(dh: np.ndarray, coords: np.ndarray, nsamp: int):
...
+    :param w: width of array
+    :param center: center
+    :param in_radius: inside radius
+    :param out_radius: outside radius
+    :return:
+    """
+
+    mask_inside = create_circular_mask(h,w,center=center,radius=in_radius)
+    mask_outside = create_circular_mask(h,w,center=center,radius=out_radius)
```
Comment (adehecq) [inline xdem/spstats.py:129]: By default, the in_radius and out_radius are equal, so will it not return an empty mask?? I would avoid this as it can be confusing.

File: tests/test_spstats.py [hunk 7]
```diff
@@ -156,6 +156,46 @@ def test_neff_estimation(self):
...
+class TestSubSampling:
+
+    def test_circular_masking(self):
+        """Test that the circular masking works as intended"""
+
+        # using default (center should be [2,2], radius 2)
+        circ = xdem.spstats.create_circular_mask(5,5)
+        circ2 = xdem.spstats.create_circular_mask(5,5,center=[2,2],radius=2)
+
+        assert np.array_equal(circ,circ2)
```
Comment (adehecq) [inline tests/test_spstats.py:168]: You could also test that it is exactly equal to the matrix you expect i.e. 0 and 1 with the correct shape. This would be a more rigorous test!

File: tests/test_spstats.py [hunk 8]
```diff
@@ -156,6 +156,46 @@ def test_neff_estimation(self):
...
+
+    def test_ring_masking(self):
+        """Test that the ring masking works as intended"""
+
+        # by default, the mask is only False (ring of size 0)
+        ring1 = xdem.spstats.create_ring_mask(5,5)
+
+        assert np.array_equal(ring1,np.zeros((5,5)))
+
+        # test rings with different inner radius
```
Comment (adehecq) [inline tests/test_spstats.py:188]: Same comment, you could create the matrix as it should look like and compare that they are the same!

File: xdem/spstats.py (hunk 2 above)
Comment (erikmannerfelt) [reply xdem/spstats.py:97]: If this is done, the type would be `tuple[int, ...]` as an array's number of dimensions is never known by the type hinting. I just mention it here because I've been stuck on it myself!

Comment (rhugonnet) [reply xdem/spstats.py:97]: Done

File: xdem/spstats.py (hunk 3 above)
Comment (rhugonnet) [reply xdem/spstats.py:93]: Actually `skimage.draw.circle` is being deprecated for `skimage.draw.disk`, which also takes ~3 lines to write because it only outputs coordinates. 
I added the skimage solution too as it's only 3 lines, it behaves slightly differently (interval not inclusive, so radius=2 won't create a circle of 2 pixels, you need to input above, 2.000001 works for instance)

File: xdem/spstats.py (hunk 4 above)
Comment (rhugonnet) [reply xdem/spstats.py:99]: Done

File: xdem/spstats.py (hunk 6 above)
Comment (rhugonnet) [reply xdem/spstats.py:129]: Now the defaults return a full circle with `inside_radius` of 0

File: tests/test_spstats.py (hunk 7 above)
Comment (rhugonnet) [reply tests/test_spstats.py:168]: Added

File: tests/test_spstats.py (hunk 8 above)
Comment (rhugonnet) [reply tests/test_spstats.py:188]: Added

File: tests/test_spstats.py (hunk 1 above)
Comment (rhugonnet) [reply tests/test_spstats.py:24]: Fixed

**Comment by rhugonnet** [comment]
@erikmannerfelt @adehecq We can probably merge this too, and add the usage of those sampling functions in another PR for more clarity

**Comment by rhugonnet** [comment]
I need an approval here to move on to the next steps :pleading_face:

**Comment by erikmannerfelt** [comment]
You need to solve the merge conflicts first, @rhugonnet!

**Review by erikmannerfelt — APPROVED** [review]

**Comment by rhugonnet** [comment]
Resolves #132
Resolves #45
